packages/zql/src/builder/builder.ts

```diff
@@ -277,9 +280,125 @@ function applyWhere(
   delegate: BuilderDelegate,
   name: string,
 ): Input {
-  return buildFilterPipeline(input, delegate, filterInput =>
-    applyFilter(filterInput, condition, delegate, name),
-  );
+  if (!conditionIncludesFlippedSubqueryAtAnyLevel(condition)) {
+    return buildFilterPipeline(input, delegate, filterInput =>
+      applyFilter(filterInput, condition, delegate, name),
+    );
+  }
+
+  return applyFilterWithFlips(input, condition, delegate, name);
+}
+
+function applyFilterWithFlips(
+  input: Input,
+  condition: Condition,
+  delegate: BuilderDelegate,
+  name: string,
+): Input {
+  let end = input;
+  assert(condition.type !== 'simple', 'Simple conditions cannot have flips');
+
+  switch (condition.type) {
+    case 'and': {
+      const [withFlipped, withoutFlipped] = partitionBranches(
+        condition.conditions,
+        conditionIncludesFlippedSubqueryAtAnyLevel,
+      );
+      if (withoutFlipped.length > 0) {
+        end = buildFilterPipeline(input, delegate, filterInput =>
+          applyAnd(
+            filterInput,
+            {
+              type: 'and',
+              conditions: withoutFlipped,
+            },
+            delegate,
+            name,
+          ),
+        );
+      }
+      assert(withFlipped.length > 0, 'Impossible to have no flips here');
+      for (const cond of withFlipped) {
+        end = applyFilterWithFlips(end, cond, delegate, name);
+      }
+      break;
+    }
+    case 'or': {
+      const [withFlipped, withoutFlipped] = partitionBranches(
+        condition.conditions,
+        conditionIncludesFlippedSubqueryAtAnyLevel,
+      );
+      assert(withFlipped.length > 0, 'Impossible to have no flips here');
+
+      const ufo = new UnionFanOut(end);
+      delegate.addEdge(end, ufo);
+      end = delegate.decorateInput(ufo, `${name}:ufo`);
+
+      const branches: Input[] = [];
+      if (withoutFlipped.length > 0) {
+        branches.push(
+          buildFilterPipeline(end, delegate, filterInput =>
+            applyOr(
+              filterInput,
+              {
+                type: 'or',
+                conditions: withoutFlipped,
+              },
+              delegate,
+              name,
+            ),
+          ),
+        );
+      }
+
+      for (const cond of withFlipped) {
+        branches.push(applyFilterWithFlips(end, cond, delegate, name));
+      }
+
+      const ufi = new UnionFanIn(ufo, branches);
+      for (const branch of branches) {
+        delegate.addEdge(branch, ufi);
+      }
+      end = delegate.decorateInput(ufi, `${name}:ufi`);
+
+      break;
+    }
+    case 'correlatedSubquery': {
+      const sq = condition.related;
+      if (!sq.flip) {
```

Comment:
Is this possible, shouldn't it be in a withoutFlipped partition and not passed here?